In [3]:
import numpy as np
import pandas as pd

df=pd.read_csv("/content/WA_Fn-UseC_-Telco-Customer-Churn.csv")

#  - Data Inspection

## Dataset: Telco Customer Churn

Data inspection is the process of examining the structure, characteristics, and contents of a dataset before performing data analysis or preprocessing.

### Objectives

- Identify the number of rows and columns
- Inspect column names
- Check data types
- Understand the dataset index
- View sample records
- Generate statistical summaries
- Identify unique values
- Identify the distribution of categorical variables
- Identify numerical, categorical, date/time, identifier, and target columns

## 1. Number of Rows

The number of rows represents the number of observations or records present in the dataset.

In this dataset, each row represents one telecom customer.

In [4]:
# Number of rows

df.shape[0]

7043

## 2. Number of Columns

The number of columns represents the number of variables or attributes present in the dataset.

Each column contains information about the customers or the target variable.

## 2. Number of Columns

The number of columns represents the number of variables or attributes present in the dataset.

Each column contains information about the customers or the target variable.

In [5]:
# Number of columns

df.shape[1]

21

## 3. Data Types

Data types describe the type of values stored in each column.

Checking data types helps identify numerical, categorical, and other types of variables and can also reveal potential data-quality issues.

In [6]:
# Check data types of all columns

df.dtypes

,0
customerID,object
gender,object
SeniorCitizen,int64
Partner,object
Dependents,object
tenure,int64
PhoneService,object
MultipleLines,object
InternetService,object
OnlineSecurity,object


## 4. Column Names

Column names identify the variables present in the dataset.

Inspecting column names helps us understand the available information and identify important columns such as identifiers and target variables.

In [7]:
# Display all column names

df.columns

Index(['customerID', 'gender', 'SeniorCitizen', 'Partner', 'Dependents',
       'tenure', 'PhoneService', 'MultipleLines', 'InternetService',
       'OnlineSecurity', 'OnlineBackup', 'DeviceProtection', 'TechSupport',
       'StreamingTV', 'StreamingMovies', 'Contract', 'PaperlessBilling',
       'PaymentMethod', 'MonthlyCharges', 'TotalCharges', 'Churn'],
      dtype='object')

## 5. Index

An index is used to identify and access individual rows in a Pandas DataFrame.

By default, Pandas creates a numerical index starting from 0 when a dataset is loaded.

In [8]:
# Display the DataFrame index

df.index

RangeIndex(start=0, stop=7043, step=1)

## 6. head()

The `head()` method displays the first five rows of a DataFrame by default.

It is useful for quickly inspecting the structure and values of the dataset.

In [9]:
# Display the first 5 rows

df.head()

,customerID,gender,SeniorCitizen,Partner,Dependents,tenure,PhoneService,MultipleLines,InternetService,OnlineSecurity,...,DeviceProtection,TechSupport,StreamingTV,StreamingMovies,Contract,PaperlessBilling,PaymentMethod,MonthlyCharges,TotalCharges,Churn
0,7590-VHVEG,Female,0,Yes,No,1,No,No phone service,DSL,No,...,No,No,No,No,Month-to-month,Yes,Electronic check,29.85,29.85,No
1,5575-GNVDE,Male,0,No,No,34,Yes,No,DSL,Yes,...,Yes,No,No,No,One year,No,Mailed check,56.95,1889.5,No
2,3668-QPYBK,Male,0,No,No,2,Yes,No,DSL,Yes,...,No,No,No,No,Month-to-month,Yes,Mailed check,53.85,108.15,Yes
3,7795-CFOCW,Male,0,No,No,45,No,No phone service,DSL,Yes,...,Yes,Yes,No,No,One year,No,Bank transfer (automatic),42.30,1840.75,No
4,9237-HQITU,Female,0,No,No,2,Yes,No,Fiber optic,No,...,No,No,No,No,Month-to-month,Yes,Electronic check,70.70,151.65,Yes


## 7. tail()

The `tail()` method displays the last five rows of a DataFrame by default.

It helps inspect the records at the end of the dataset.

In [10]:
# Display the last 5 rows

df.tail()

,customerID,gender,SeniorCitizen,Partner,Dependents,tenure,PhoneService,MultipleLines,InternetService,OnlineSecurity,...,DeviceProtection,TechSupport,StreamingTV,StreamingMovies,Contract,PaperlessBilling,PaymentMethod,MonthlyCharges,TotalCharges,Churn
7038,6840-RESVB,Male,0,Yes,Yes,24,Yes,Yes,DSL,Yes,...,Yes,Yes,Yes,Yes,One year,Yes,Mailed check,84.80,1990.5,No
7039,2234-XADUH,Female,0,Yes,Yes,72,Yes,Yes,Fiber optic,No,...,Yes,No,Yes,Yes,One year,Yes,Credit card (automatic),103.20,7362.9,No
7040,4801-JZAZL,Female,0,Yes,Yes,11,No,No phone service,DSL,Yes,...,No,No,No,No,Month-to-month,Yes,Electronic check,29.60,346.45,No
7041,8361-LTMKD,Male,1,Yes,No,4,Yes,Yes,Fiber optic,No,...,No,No,No,No,Month-to-month,Yes,Mailed check,74.40,306.6,Yes
7042,3186-AJIEK,Male,0,No,No,66,Yes,No,Fiber optic,Yes,...,Yes,Yes,Yes,Yes,Two year,Yes,Bank transfer (automatic),105.65,6844.5,No


## 8. sample()

The `sample()` method randomly selects rows from a DataFrame.

It is useful for inspecting random observations instead of only looking at the beginning or end of the dataset.

In [11]:
# Display 5 random rows

df.sample(5)

,customerID,gender,SeniorCitizen,Partner,Dependents,tenure,PhoneService,MultipleLines,InternetService,OnlineSecurity,...,DeviceProtection,TechSupport,StreamingTV,StreamingMovies,Contract,PaperlessBilling,PaymentMethod,MonthlyCharges,TotalCharges,Churn
3849,5089-IFSDP,Female,0,Yes,No,58,Yes,Yes,Fiber optic,Yes,...,Yes,No,Yes,Yes,Two year,Yes,Bank transfer (automatic),109.45,6144.55,Yes
6456,9174-IHETN,Female,0,No,Yes,50,Yes,No,Fiber optic,No,...,No,No,No,No,Two year,No,Credit card (automatic),70.35,3533.6,No
4216,2990-IAJSV,Male,0,No,No,72,Yes,Yes,DSL,Yes,...,Yes,Yes,Yes,Yes,Two year,No,Bank transfer (automatic),92.00,6632.75,No
6807,2030-BTZRO,Male,0,Yes,Yes,6,Yes,No,No,No internet service,...,No internet service,No internet service,No internet service,No internet service,Two year,Yes,Credit card (automatic),20.40,107.6,No
218,2040-LDIWQ,Male,0,Yes,Yes,65,Yes,Yes,DSL,No,...,Yes,Yes,Yes,Yes,Two year,Yes,Bank transfer (automatic),84.20,5324.5,No


## 9. shape

The `shape` attribute returns the dimensions of a DataFrame in the form:

`(number_of_rows, number_of_columns)`

For this dataset, the shape tells us how many customer records and variables are present.

In [12]:
# Check the dimensions of the dataset

df.shape

(7043, 21)

## 10. info()

The `info()` method provides a concise summary of the DataFrame.

It displays:

- Number of entries
- Column names
- Number of non-null values
- Data types
- Memory usage

It is useful for identifying data types and possible missing values.

In [13]:
# Display information about the dataset

df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 21 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerID        7043 non-null   object 
 1   gender            7043 non-null   object 
 2   SeniorCitizen     7043 non-null   int64  
 3   Partner           7043 non-null   object 
 4   Dependents        7043 non-null   object 
 5   tenure            7043 non-null   int64  
 6   PhoneService      7043 non-null   object 
 7   MultipleLines     7043 non-null   object 
 8   InternetService   7043 non-null   object 
 9   OnlineSecurity    7043 non-null   object 
 10  OnlineBackup      7043 non-null   object 
 11  DeviceProtection  7043 non-null   object 
 12  TechSupport       7043 non-null   object 
 13  StreamingTV       7043 non-null   object 
 14  StreamingMovies   7043 non-null   object 
 15  Contract          7043 non-null   object 
 16  PaperlessBilling  7043 non-null   object 


### Observation

The dataset contains 7,043 entries and 21 columns.

Most columns have `object` data type, while `SeniorCitizen` and `tenure` are `int64`, and `MonthlyCharges` is `float64`.

`TotalCharges` is stored as `object` even though it represents numerical information. This indicates a potential data-quality issue that should be investigated during data cleaning.

## 11. describe()

The `describe()` method generates descriptive statistics for numerical columns by default.

It provides:

- Count
- Mean
- Standard deviation
- Minimum
- 25th percentile
- Median (50th percentile)
- 75th percentile
- Maximum

These statistics help us understand the distribution and general characteristics of numerical variables.

In [14]:
# Generate descriptive statistics for numerical columns

df.describe()

,SeniorCitizen,tenure,MonthlyCharges
count,7043.000000,7043.000000,7043.000000
mean,0.162147,32.371149,64.761692
std,0.368612,24.559481,30.090047
min,0.000000,0.000000,18.250000
25%,0.000000,9.000000,35.500000
50%,0.000000,29.000000,70.350000
75%,0.000000,55.000000,89.850000
max,1.000000,72.000000,118.750000


### Observation

The numerical summary provides information about variables such as `SeniorCitizen`, `tenure`, and `MonthlyCharges`.

It helps identify the central tendency, spread, and range of numerical variables and can also help detect unusual values that may require further investigation.

## 12. dtypes

The `dtypes` attribute returns the data type of every column in the DataFrame.

It helps us identify which columns are numerical and which columns are stored as categorical or object data.

In [15]:
# Display data types of all columns

df.dtypes

,0
customerID,object
gender,object
SeniorCitizen,int64
Partner,object
Dependents,object
tenure,int64
PhoneService,object
MultipleLines,object
InternetService,object
OnlineSecurity,object


### Observation

The dataset contains:

- `int64` columns: `SeniorCitizen`, `tenure`
- `float64` column: `MonthlyCharges`
- `object` columns: categorical columns, `customerID`, `Churn`, and `TotalCharges`

`TotalCharges` requires further investigation because it contains numerical information but is stored as `object`.

## 13. nunique()

The `nunique()` method returns the number of unique values in each column.

It is useful for understanding the number of distinct values in a feature and can help identify categorical variables, binary variables, and identifier columns.

In [16]:
# Count unique values in each column

df.nunique()

,0
customerID,7043
gender,2
SeniorCitizen,2
Partner,2
Dependents,2
tenure,73
PhoneService,2
MultipleLines,3
InternetService,3
OnlineSecurity,3


### Observation

The `nunique()` result helps identify the nature of different columns.

For example:

- `customerID` has a very large number of unique values because it identifies individual customers.
- `Churn` has 2 unique values.
- `SeniorCitizen` has 2 unique values.
- Categorical columns have a limited number of unique categories.

This information is useful when deciding how different columns should be analyzed and processed.

## 14. value_counts()

The `value_counts()` method counts how many times each unique value occurs in a column.

It is mainly useful for analyzing categorical variables and understanding the distribution of their categories.

In [17]:
# Count the values of the target variable

df["Churn"].value_counts()

,count
Churn,
No,5174
Yes,1869


### Observation

The `Churn` column contains two categories:

- `No`
- `Yes`

The value counts show how many customers belong to each category.

This is important because it helps us understand the distribution of the target variable and identify whether the classes are balanced or imbalanced.

In [18]:
# Calculate the percentage distribution of the target variable

df["Churn"].value_counts(normalize=True) * 100

,proportion
Churn,
No,73.463013
Yes,26.536987


### Inspecting Other Categorical Variables

`value_counts()` can also be used to understand the distribution of individual categorical features.

In [19]:
# Distribution of Contract categories

df["Contract"].value_counts()

,count
Contract,
Month-to-month,3875
Two year,1695
One year,1473


In [20]:
# Distribution of InternetService categories

df["InternetService"].value_counts()

,count
InternetService,
Fiber optic,3096
DSL,2421
No,1526


## 15. Column Classification

After inspecting the dataset, the columns can be classified into different groups:

- Numerical columns
- Categorical columns
- Date/time columns
- Identifier columns
- Target column

This classification helps determine the appropriate analysis and preprocessing techniques for each type of variable.

In [21]:
# Numerical columns

numerical_columns = df.select_dtypes(include=np.number).columns.tolist()

# Categorical/object columns

categorical_columns = df.select_dtypes(include="object").columns.tolist()

# Identifier column

identifier_columns = ["customerID"]

# Target column

target_column = ["Churn"]

print("Numerical Columns:")
print(numerical_columns)

print("\nCategorical/Object Columns:")
print(categorical_columns)

print("\nIdentifier Column:")
print(identifier_columns)

print("\nTarget Column:")
print(target_column)

Numerical Columns:
['SeniorCitizen', 'tenure', 'MonthlyCharges']

Categorical/Object Columns:
['customerID', 'gender', 'Partner', 'Dependents', 'PhoneService', 'MultipleLines', 'InternetService', 'OnlineSecurity', 'OnlineBackup', 'DeviceProtection', 'TechSupport', 'StreamingTV', 'StreamingMovies', 'Contract', 'PaperlessBilling', 'PaymentMethod', 'TotalCharges', 'Churn']

Identifier Column:
['customerID']

Target Column:
['Churn']


## Data Inspection Summary

The dataset contains 7,043 rows and 21 columns.

The inspection identified:

- Numerical columns: `SeniorCitizen`, `tenure`, `MonthlyCharges`
- Object columns: categorical variables, `customerID`, `TotalCharges`, and `Churn`
- Identifier column: `customerID`
- Target column: `Churn`
- Date/time columns: None

`TotalCharges` requires further investigation because it represents numerical information but is stored as an object data type.

The target variable `Churn` contains two classes, `Yes` and `No`.